plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/BaseEqualsVisitor.java

```diff
@@ -53,19 +67,156 @@ public void visitMethodReferenceExpression(PsiMethodReferenceExpression expressi
+    PsiType returnParameter = findPredicateParameter(upperPredicate);
+    if (returnParameter != null) {
+      return returnParameter;
     }
+
+    if (parent instanceof PsiMethodCallExpression) {
+      final PsiMethodCallExpression callExpresion = (PsiMethodCallExpression)parent;
+      if (PREDICATE_TEST.test(callExpresion)) {
+        final PsiExpression[] argumentExpressions = callExpresion.getArgumentList().getExpressions();
+        if (argumentExpressions.length != 1) {
+          return null;
+        }
+        PsiExpression argument = PsiUtil.skipParenthesizedExprDown(argumentExpressions[0]);
+        return getType(argument);
+      }
+    }
+    else if (parent instanceof PsiExpressionList) {
+      return findPredicateTypeFromExpressionList(upperPredicate, (PsiExpressionList)parent);
+    }
+    return null;
+  }
+
+  @Nullable
+  private static PsiType findPredicateTypeFromExpressionList(PsiExpression upperPredicate, PsiExpressionList expressionList) {
+    Integer index = findParameterIndex(expressionList, upperPredicate);
+    if (index == null || expressionList.getExpressionCount() <= index) {
+      return null;
+    }
+
+    if (!(expressionList.getParent() instanceof PsiMethodCallExpression)) {
+      return null;
+    }
+
+    final PsiMethodCallExpression expectedCall = (PsiMethodCallExpression)expressionList.getParent();
+    PsiMethod method = expectedCall.resolveMethod();
+    if (method == null) {
+      return null;
+    }
+    PsiType[] arguments = method.getSignature(expectedCall.resolveMethodGenerics().getSubstitutor())
+      .getParameterTypes();
+
+    if (arguments.length <= index) {
+      return null;
+    }
+    PsiType type = arguments[index];
+    if (!(type instanceof PsiClassType)) {
+      return null;
+    }
+    PsiClassType classType = (PsiClassType)type;
+    if (classType.getParameterCount() != 1 || !PsiTypesUtil.classNameEquals(classType, CommonClassNames.JAVA_UTIL_FUNCTION_PREDICATE)) {
+      return null;
+    }
+    PsiType parameter = classType.getParameters()[0];
+    //Object can be cast to any
+    if (parameter == null || parameter.equalsToText(CommonClassNames.JAVA_LANG_OBJECT)) {
+      return null;
+    }
+    if (parameter instanceof PsiWildcardType) {
+      PsiWildcardType psiWildcardType = (PsiWildcardType)parameter;
+      parameter = psiWildcardType.isSuper() ? psiWildcardType.getSuperBound() : parameter;
+    }
+    return parameter;
+  }
+
+  @Nullable
+  private static Integer findParameterIndex(PsiExpressionList list, PsiExpression predicate) {
+    for (int i = 0; i < list.getExpressionCount(); i++) {
+      final PsiExpression expression = list.getExpressions()[i];
+      if (predicate.equals(PsiUtil.skipParenthesizedExprDown(expression))) {
+        return i;
+      }
+    }
+    return null;
+  }
+
+  @Nullable
+  private static PsiType findPredicateParameter(PsiExpression predicate) {
```

Comment:
Why `findPredicateParameter` if it returns the type?

Reply:
Thank you. My fault, don't pay attention. Rename method